diff --git a/CMakeLists.txt b/CMakeLists.txt
index eb89404..038b896 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,7 +15,7 @@ set(SUPPORTED_TARGETS "raspi64" "pbpro" "rockpro64" "virt")
 set(TARGET "raspi64" CACHE STRING "One of target machines: ${SUPPORTED_TARGETS}")
 set_property(CACHE TARGET PROPERTY STRINGS ${SUPPORTED_TARGETS})
 
-set(SUPPORTED_VARIANTS "none" "pistorm" "pistorm32lite")
+set(SUPPORTED_VARIANTS "none" "pistorm" "pistorm32lite" "bellatrix")
 set(VARIANT "none" CACHE STRING "Compilation variant: ${SUPPORTED_VARIANTS}")
 set_property(CACHE VARIANT PROPERTY STRINGS ${SUPPORTED_VARIANTS})
 
@@ -167,14 +167,18 @@ if (${VARIANT} IN_LIST SUPPORTED_VARIANTS)
         else()
             message(FATAL_ERROR "PiStorm variants are supported on raspi targets, only.")
         endif()
+    elseif(${VARIANT} STREQUAL "bellatrix")
+        include(${CMAKE_SOURCE_DIR}/../cmake/bellatrix-variant.cmake)
     endif()
-    
+
 else()
     message(FATAL_ERROR "Wrong variant selected: ${VARIANT}")
 endif()
 
+
 add_link_options(-Wl,--build-id -nostdlib -nostartfiles -static)
 add_compile_options(-mbig-endian -fno-exceptions -ffreestanding -Wall -Wextra -Werror)
+add_compile_options(-fno-stack-protector -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0)
 
 list(APPEND ARCH_FILES
     src/aarch64/start.c
@@ -209,7 +213,9 @@ add_subdirectory(external)
 target_link_libraries(Emu68.elf capstone-static libdeflate_static)
 target_include_directories(Emu68.elf PRIVATE ${CMAKE_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/external/capstone/include)
 target_include_directories(Emu68.elf PRIVATE ${CMAKE_BINARY_DIR}/include)
-target_compile_definitions(Emu68.elf PRIVATE)
+if(${VARIANT} STREQUAL "bellatrix")
+    include(${CMAKE_SOURCE_DIR}/../cmake/bellatrix-target.cmake)
+endif()
 
 configure_file(include/version.h.in include/version.h @ONLY)
 set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${CMAKE_SOURCE_DIR}/.git/index)
