From 7bb40f5a64be826112cbc85e2ede909e837e2644 Mon Sep 17 00:00:00 2001
From: Jaime Dias <jaimejosediasnt@gmail.com>
Date: Tue, 14 Apr 2026 05:11:27 -0300
Subject: [PATCH 1/2] Add bellatrix to SUPPORTED_VARIANTS and configure sources

---
 CMakeLists.txt | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index eb89404..303d1a0 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,7 +15,7 @@ set(SUPPORTED_TARGETS "raspi64" "pbpro" "rockpro64" "virt")
 set(TARGET "raspi64" CACHE STRING "One of target machines: ${SUPPORTED_TARGETS}")
 set_property(CACHE TARGET PROPERTY STRINGS ${SUPPORTED_TARGETS})
 
-set(SUPPORTED_VARIANTS "none" "pistorm" "pistorm32lite")
+set(SUPPORTED_VARIANTS "none" "pistorm" "pistorm32lite" "bellatrix")
 set(VARIANT "none" CACHE STRING "Compilation variant: ${SUPPORTED_VARIANTS}")
 set_property(CACHE VARIANT PROPERTY STRINGS ${SUPPORTED_VARIANTS})
 
@@ -167,8 +167,25 @@ if (${VARIANT} IN_LIST SUPPORTED_VARIANTS)
         else()
             message(FATAL_ERROR "PiStorm variants are supported on raspi targets, only.")
         endif()
+    elseif(${VARIANT} STREQUAL "bellatrix")
+        if(${TARGET} STREQUAL "raspi64")
+            add_compile_definitions(BELLATRIX)
+            include_directories(${CMAKE_SOURCE_DIR}/../src)
+            include_directories(${CMAKE_SOURCE_DIR}/../src/cpu)
+            include_directories(${CMAKE_SOURCE_DIR}/../src/host)
+            list(APPEND BASE_FILES
+                ${CMAKE_SOURCE_DIR}/../src/cpu/bellatrix.c
+                ${CMAKE_SOURCE_DIR}/../src/core/btrace.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/cia/cia.c
+                ${CMAKE_SOURCE_DIR}/../src/host/raspi3/pal_debug.c
+                ${CMAKE_SOURCE_DIR}/../src/host/raspi3/pal_ipl.c
+                ${CMAKE_SOURCE_DIR}/../src/host/raspi3/pal_core.c
+            )
+        else()
+            message(FATAL_ERROR "Bellatrix variant is supported on raspi64 target only.")
+        endif()
     endif()
-    
+
 else()
     message(FATAL_ERROR "Wrong variant selected: ${VARIANT}")
 endif()
-- 
2.34.1

