--- /tmp/CMakeLists.orig	2026-04-20 17:18:28.206431937 -0300
+++ CMakeLists.txt	2026-04-20 17:14:00.906425519 -0300
@@ -15,7 +15,7 @@
 set(TARGET "raspi64" CACHE STRING "One of target machines: ${SUPPORTED_TARGETS}")
 set_property(CACHE TARGET PROPERTY STRINGS ${SUPPORTED_TARGETS})
 
-set(SUPPORTED_VARIANTS "none" "pistorm" "pistorm32lite")
+set(SUPPORTED_VARIANTS "none" "pistorm" "pistorm32lite" "bellatrix")
 set(VARIANT "none" CACHE STRING "Compilation variant: ${SUPPORTED_VARIANTS}")
 set_property(CACHE VARIANT PROPERTY STRINGS ${SUPPORTED_VARIANTS})
 
@@ -167,8 +167,32 @@
         else()
             message(FATAL_ERROR "PiStorm variants are supported on raspi targets, only.")
         endif()
+    elseif(${VARIANT} STREQUAL "bellatrix")
+        if(${TARGET} STREQUAL "raspi64")
+            add_compile_definitions(BELLATRIX)
+            include_directories(${CMAKE_SOURCE_DIR}/../src)
+            include_directories(${CMAKE_SOURCE_DIR}/../src/cpu)
+            include_directories(${CMAKE_SOURCE_DIR}/../src/host)
+            list(APPEND BASE_FILES
+                ${CMAKE_SOURCE_DIR}/../src/cpu/bellatrix.c
+                ${CMAKE_SOURCE_DIR}/../src/core/machine.c
+                ${CMAKE_SOURCE_DIR}/../src/core/btrace.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/cia/cia.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/agnus/agnus.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/agnus/blitter.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/agnus/copper.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/denise/denise.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/paula/paula.c
+                ${CMAKE_SOURCE_DIR}/../src/chipset/agnus/dma.c
+                ${CMAKE_SOURCE_DIR}/../src/host/raspi3/pal_debug.c
+                ${CMAKE_SOURCE_DIR}/../src/host/raspi3/pal_ipl.c
+                ${CMAKE_SOURCE_DIR}/../src/host/raspi3/pal_core.c
+            )
+        else()
+            message(FATAL_ERROR "Bellatrix variant is supported on raspi64 target only.")
+        endif()
     endif()
-    
+
 else()
     message(FATAL_ERROR "Wrong variant selected: ${VARIANT}")
 endif()
