--- a/src/aarch64/start.c
+++ b/src/aarch64/start.c
@@ -28,6 +28,9 @@
 #include "version.h"
 #include "cache.h"
 #include "sponsoring.h"
+#ifdef BELLATRIX
+#include "bellatrix.h"
+#endif
 
 void _start();
 void _boot();
@@ -1189,7 +1192,7 @@
         InitFunctions++;
     }
 
-#ifndef PISTORM
+#if !defined(PISTORM) && !defined(BELLATRIX)
     if (initramfs_loc != NULL && initramfs_size != 0)
     {
         void *image_start, *image_end;
@@ -1305,6 +1308,7 @@
 
 #else
 
+#ifdef PISTORM
     if (rom_copy != 0)
     {
         kprintf("[BOOT] %dk ROM copy requested\n", rom_copy);
@@ -1351,7 +1355,8 @@
             mmu_map(0xf80000, 0xe00000, 524288, MMU_ACCESS | MMU_ISHARE | MMU_ALLOW_EL0 | MMU_READ_ONLY | MMU_ATTR_CACHED, 0);
         }
     }
-    else if (initramfs_loc != NULL && initramfs_size != 0)
+#endif /* PISTORM */
+    if (initramfs_loc != NULL && initramfs_size != 0)
     {
         extern uint32_t rom_mapped;
 
@@ -1499,7 +1504,11 @@
     if (fast_page0) {
         mmu_map(0xf80000, 0x0, 4096, MMU_ACCESS | MMU_ISHARE | MMU_ALLOW_EL0 | MMU_READ_ONLY | MMU_ATTR_CACHED, 0);
     }
-    
+
+#ifdef BELLATRIX
+    bellatrix_init();
+#endif
+
     M68K_StartEmu(0, NULL);
 
 #endif
--- a/src/aarch64/vectors.c
+++ b/src/aarch64/vectors.c
@@ -661,6 +661,34 @@
     return 1;
 }
 
+#elif defined(BELLATRIX)
+
+#include "bellatrix.h"
+
+int SYSWriteValToAddr(uint64_t value, uint64_t value2, int size, uint64_t far)
+{
+    D(kprintf("[JIT:SYS] SYSWriteValToAddr(0x%x, %d, %p)\n", value, size, far));
+    (void)value2;
+
+    if ((far >> 32) == 1 || (far >> 32) == 0xffffffff)
+        far &= 0xffffffff;
+
+    bellatrix_bus_access((uint32_t)far, (uint32_t)value, size, BUS_WRITE);
+    return 1;
+}
+
+int SYSReadValFromAddr(uint64_t *value, uint64_t *value2, int size, uint64_t far)
+{
+    D(kprintf("[JIT:SYS] SYSReadValFromAddr(%d, %p)\n", size, far));
+
+    if ((far >> 32) == 1 || (far >> 32) == 0xffffffff)
+        far &= 0xffffffff;
+
+    *value = bellatrix_bus_access((uint32_t)far, 0, size, BUS_READ);
+    if (value2) *value2 = 0;
+    return 1;
+}
+
 #else
 
 int SYSWriteValToAddr(uint64_t value, uint64_t value2, int size, uint64_t far)
