diff --git a/src/aarch64/vectors.c b/src/aarch64/vectors.c
index 329d470..b186318 100644
--- a/src/aarch64/vectors.c
+++ b/src/aarch64/vectors.c
@@ -661,6 +661,10 @@ int SYSReadValFromAddr(uint64_t *value, uint64_t *value2, int size, uint64_t far
     return 1;
 }
 
+#elif defined(BELLATRIX)
+
+#include "cpu/emu68/vectors.inc"
+
 #else
 
 int SYSWriteValToAddr(uint64_t value, uint64_t value2, int size, uint64_t far)
