diff --git a/src/aarch64/mmu.c b/src/aarch64/mmu.c
index 3c89731..8447bd0 100644
--- a/src/aarch64/mmu.c
+++ b/src/aarch64/mmu.c
@@ -550,7 +550,78 @@ void mmu_map(uintptr_t phys, uintptr_t virt, uintptr_t length, uint32_t attr_low
 
 void mmu_unmap(uintptr_t virt, uintptr_t length)
 {
-    (void)virt;
-    (void)length;
     DMAP(kprintf("mmu_unmap(%p, %x)\n", virt, length));
+
+    while (length >= 4096)
+    {
+        struct mmu_page *l1;
+        struct mmu_page *l2;
+        struct mmu_page *l3;
+        uint64_t entry;
+        unsigned idx_l1 = (virt >> 30) & 0x1ff;
+        unsigned idx_l2 = (virt >> 21) & 0x1ff;
+        unsigned idx_l3 = (virt >> 12) & 0x1ff;
+
+        if (virt & 0xffff000000000000) {
+            asm volatile("mrs %0, TTBR1_EL1":"=r"(l1));
+        } else {
+            asm volatile("mrs %0, TTBR0_EL1":"=r"(l1));
+        }
+        l1 = (struct mmu_page *)((uintptr_t)l1 + PHYS_VIRT_OFFSET);
+
+        entry = l1->mp_entries[idx_l1];
+        if ((entry & 3) == 1)
+        {
+            l2 = get_4k_page();
+            for (unsigned i = 0; i < 512; ++i)
+                l2->mp_entries[i] = (entry & 0x7fc0000fff) +
+                                    ((uint64_t)i << 21);
+            l1->mp_entries[idx_l1] =
+                3 | ((uintptr_t)l2 - PHYS_VIRT_OFFSET);
+            mirror_page(virt);
+        }
+        else if ((entry & 3) == 3)
+        {
+            l2 = (struct mmu_page *)((entry & 0x7ffffff000ULL) +
+                                     PHYS_VIRT_OFFSET);
+        }
+        else
+        {
+            virt += 4096;
+            length -= 4096;
+            continue;
+        }
+
+        entry = l2->mp_entries[idx_l2];
+        if ((entry & 3) == 1)
+        {
+            l3 = get_4k_page();
+            for (unsigned i = 0; i < 512; ++i)
+                l3->mp_entries[i] =
+                    3 | ((entry & 0x7fffe00fff) + ((uint64_t)i << 12));
+            l2->mp_entries[idx_l2] =
+                3 | ((uintptr_t)l3 - PHYS_VIRT_OFFSET);
+        }
+        else if ((entry & 3) == 3)
+        {
+            l3 = (struct mmu_page *)((entry & 0x7ffffff000ULL) +
+                                     PHYS_VIRT_OFFSET);
+        }
+        else
+        {
+            virt += 4096;
+            length -= 4096;
+            continue;
+        }
+
+        l3->mp_entries[idx_l3] = 0;
+        virt += 4096;
+        length -= 4096;
+    }
+
+    asm volatile(
+"       dsb     ish                 \n"
+"       tlbi    VMALLE1IS           \n"
+"       dsb     sy                  \n"
+"       isb                         \n");
 }
 
