diff --git a/src/M68k_LINEE.c b/src/M68k_LINEE.c
index ec68621..f6dce91 100644
--- a/src/M68k_LINEE.c
+++ b/src/M68k_LINEE.c
@@ -24,10 +24,10 @@ static uint32_t *EMIT_ASL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_pt
 
     /* Pre-decrement mode */
     if ((opcode & 0x38) == 0x20) {
-        *ptr++ = ldrsh_offset_preindex(dest, tmp, -2);
+        EMU68_GUEST_PRELOAD(ptr, dest, tmp, 2, 1, -2);
     }
     else {
-        *ptr++ = ldrsh_offset(dest, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, dest, tmp, 2, 1);
     }
 
     if (update_mask & (SR_C | SR_X)) {
@@ -49,10 +49,10 @@ static uint32_t *EMIT_ASL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_pt
     }
 
     if ((opcode & 0x38) == 0x18) {
-        *ptr++ = strh_offset_postindex(dest, tmp, 2);
+        EMU68_GUEST_POSTSTORE(ptr, dest, tmp, 2, 2);
     }
     else {
-        *ptr++ = strh_offset(dest, tmp, 0);
+        EMU68_GUEST_STORE(ptr, dest, tmp, 2);
     }
 
     ptr = EMIT_AdvancePC(ptr, 2 * (ext_words + 1));
@@ -116,10 +116,10 @@ static uint32_t *EMIT_LSL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_pt
 
     /* Pre-decrement mode */
     if ((opcode & 0x38) == 0x20) {
-        *ptr++ = ldrh_offset_preindex(dest, tmp, -2);
+        EMU68_GUEST_PRELOAD(ptr, dest, tmp, 2, 0, -2);
     }
     else {
-        *ptr++ = ldrh_offset(dest, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, dest, tmp, 2, 0);
     }
 
     if (update_mask & (SR_C | SR_X)) {
@@ -141,10 +141,10 @@ static uint32_t *EMIT_LSL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_pt
     }
 
     if ((opcode & 0x38) == 0x18) {
-        *ptr++ = strh_offset_postindex(dest, tmp, 2);
+        EMU68_GUEST_POSTSTORE(ptr, dest, tmp, 2, 2);
     }
     else {
-        *ptr++ = strh_offset(dest, tmp, 0);
+        EMU68_GUEST_STORE(ptr, dest, tmp, 2);
     }
         
     ptr = EMIT_AdvancePC(ptr, 2 * (ext_words + 1));
@@ -201,10 +201,10 @@ static uint32_t *EMIT_ROXL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_p
     uint8_t cc = RA_ModifyCC(&ptr);
 
     if ((opcode & 0x38) == 0x20) {
-        *ptr++ = ldrh_offset_preindex(dest, tmp, -2);
+        EMU68_GUEST_PRELOAD(ptr, dest, tmp, 2, 0, -2);
     }
     else {
-        *ptr++ = ldrh_offset(dest, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, dest, tmp, 2, 0);
     }
 
     /* Test X flag, push the flag value into tmp register */
@@ -221,10 +221,10 @@ static uint32_t *EMIT_ROXL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_p
     }
 
     if ((opcode & 0x38) == 0x18) {
-        *ptr++ = strh_offset_postindex(dest, tmp, 2);
+        EMU68_GUEST_POSTSTORE(ptr, dest, tmp, 2, 2);
     }
     else {
-        *ptr++ = strh_offset(dest, tmp, 0);
+        EMU68_GUEST_STORE(ptr, dest, tmp, 2);
     }
 
     ptr = EMIT_AdvancePC(ptr, 2 * (ext_words + 1));
@@ -279,10 +279,10 @@ static uint32_t *EMIT_ROL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_pt
     ptr = EMIT_LoadFromEffectiveAddress(ptr, 0, &dest, opcode & 0x3f, *m68k_ptr, &ext_words, 1, NULL);
 
     if ((opcode & 0x38) == 0x20) {
-        *ptr++ = ldrh_offset_preindex(dest, tmp, -2);
+        EMU68_GUEST_PRELOAD(ptr, dest, tmp, 2, 0, -2);
     }
     else {
-        *ptr++ = ldrh_offset(dest, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, dest, tmp, 2, 0);
     }
     *ptr++ = bfi(tmp, tmp, 16, 16);
 
@@ -296,10 +296,10 @@ static uint32_t *EMIT_ROL_mem(uint32_t *ptr, uint16_t opcode, uint16_t **m68k_pt
     }
 
     if ((opcode & 0x38) == 0x18) {
-        *ptr++ = strh_offset_postindex(dest, tmp, 2);
+        EMU68_GUEST_POSTSTORE(ptr, dest, tmp, 2, 2);
     }
     else {
-        *ptr++ = strh_offset(dest, tmp, 0);
+        EMU68_GUEST_STORE(ptr, dest, tmp, 2);
     }
 
     ptr = EMIT_AdvancePC(ptr, 2 * (ext_words + 1));
@@ -1625,28 +1625,28 @@ static inline uint32_t *EMIT_BFxxx_II(uint32_t *ptr, uint8_t base, enum BF_OP op
     /* IF bit offset + width <= 8, fetch a byte */
     if ((bit_offset + width) <= 8)
     {
-        *ptr++ = ldurb_offset(base, data_reg, base_offset);
+        EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 1, 0);
         fetched_size = 1;
         data_offset = 56;
     }
     /* bit offset + width <= 16: fetch a word */
     else if ((bit_offset + width) <= 16)
     {
-        *ptr++ = ldurh_offset(base, data_reg, base_offset);
+        EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 2, 0);
         fetched_size = 2;
         data_offset = 48;
     }
     /* bit offset + width <= 32: fetch a long */
     else if ((bit_offset + width) <= 32)
     {
-        *ptr++ = ldur_offset(base, data_reg, base_offset);
+        EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 4, 0);
         fetched_size = 4;
         data_offset = 32;
     }
     /* Worst case otherwise - fetch 64bit */
     else
     {
-        *ptr++ = ldur64_offset(base, data_reg, base_offset);
+        EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 8, 0);
         fetched_size = 8;
     }
 
@@ -1753,16 +1753,20 @@ static inline uint32_t *EMIT_BFxxx_II(uint32_t *ptr, uint8_t base, enum BF_OP op
             switch (fetched_size)
             {
                 case 1:
-                    *ptr++ = sturb_offset(base, data_reg, base_offset);
+                    EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset,
+                                             data_reg, 1);
                     break;
                 case 2:
-                    *ptr++ = sturh_offset(base, data_reg, base_offset);
+                    EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset,
+                                             data_reg, 2);
                     break;
                 case 4:
-                    *ptr++ = stur_offset(base, data_reg, base_offset);
+                    EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset,
+                                             data_reg, 4);
                     break;
                 case 8:
-                    *ptr++ = stur64_offset(base, data_reg, base_offset);
+                    EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset,
+                                             data_reg, 8);
                     break;
             }
         }
@@ -1793,26 +1797,36 @@ static inline uint32_t *EMIT_BFxxx_IR(uint32_t *ptr, uint8_t base, enum BF_OP op
     /* Move mask to the topmost bits of 64-bit mask_reg */
     *ptr++ = rbit64(mask_reg, mask_reg);
 
-    /* Fetch the data */
-    /* Width == 1? Fetch byte */
+    /* Fetch the smallest container that covers the runtime width. */
     *ptr++ = cmp_immed(width_reg, 1);
-    *ptr++ = b_cc(A64_CC_NE, 4);
-    *ptr++ = ldurb_offset(base, data_reg, base_offset);
+    uint32_t *not_byte = ptr++;
+    EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 1, 0);
     *ptr++ = ror64(data_reg, data_reg, 8);
-    *ptr++ = b(12);
-    /* Width <= 8? Fetch half word */
+    uint32_t *byte_done = ptr++;
+
+    uint32_t *half_label = ptr;
     *ptr++ = cmp_immed(width_reg, 8);
-    *ptr++ = b_cc(A64_CC_GT, 4);
-    *ptr++ = ldurh_offset(base, data_reg, base_offset);
+    uint32_t *not_half = ptr++;
+    EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 2, 0);
     *ptr++ = ror64(data_reg, data_reg, 16);
-    *ptr++ = b(7);
-    /* Width <= 24? Fetch long word */
+    uint32_t *half_done = ptr++;
+
+    uint32_t *long_label = ptr;
     *ptr++ = cmp_immed(width_reg, 24);
-    *ptr++ = b_cc(A64_CC_GT, 4);
-    *ptr++ = ldur_offset(base, data_reg, base_offset);
+    uint32_t *not_long = ptr++;
+    EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 4, 0);
     *ptr++ = ror64(data_reg, data_reg, 32);
-    *ptr++ = b(2);
-    *ptr++ = ldur64_offset(base, data_reg, base_offset);
+    uint32_t *long_done = ptr++;
+
+    uint32_t *quad_label = ptr;
+    EMU68_GUEST_LOAD_OFFSET(ptr, base, base_offset, data_reg, 8, 0);
+    uint32_t *load_done = ptr;
+    *not_byte = b_cc(A64_CC_NE, half_label - not_byte);
+    *not_half = b_cc(A64_CC_GT, long_label - not_half);
+    *not_long = b_cc(A64_CC_GT, quad_label - not_long);
+    *byte_done = b(load_done - byte_done);
+    *half_done = b(load_done - half_done);
+    *long_done = b(load_done - long_done);
 
     /* In case of INS, prepare the source data accordingly */
     if (op == OP_INS)
@@ -1920,26 +1934,39 @@ static inline uint32_t *EMIT_BFxxx_IR(uint32_t *ptr, uint8_t base, enum BF_OP op
 
         if (op != OP_EXTS && op != OP_EXTU && op != OP_FFO)
         {
-            /* Store the data back */
-            /* Width == 1? Fetch byte */
+            /* Store the selected container back. */
             *ptr++ = cmp_immed(width_reg, 1);
-            *ptr++ = b_cc(A64_CC_NE, 4);
+            uint32_t *store_not_byte = ptr++;
             *ptr++ = ror64(data_reg, data_reg, 64 - 8);
-            *ptr++ = sturb_offset(base, data_reg, base_offset);
-            *ptr++ = b(12);
-            /* Width <= 8? Fetch half word */
+            EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset, data_reg, 1);
+            uint32_t *store_byte_done = ptr++;
+
+            uint32_t *store_half_label = ptr;
             *ptr++ = cmp_immed(width_reg, 8);
-            *ptr++ = b_cc(A64_CC_GT, 4);
+            uint32_t *store_not_half = ptr++;
             *ptr++ = ror64(data_reg, data_reg, 64 - 16);
-            *ptr++ = sturh_offset(base, data_reg, base_offset);
-            *ptr++ = b(7);
-            /* Width <= 24? Fetch long word */
+            EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset, data_reg, 2);
+            uint32_t *store_half_done = ptr++;
+
+            uint32_t *store_long_label = ptr;
             *ptr++ = cmp_immed(width_reg, 24);
-            *ptr++ = b_cc(A64_CC_GT, 4);
+            uint32_t *store_not_long = ptr++;
             *ptr++ = ror64(data_reg, data_reg, 32);
-            *ptr++ = stur_offset(base, data_reg, base_offset);
-            *ptr++ = b(2);
-            *ptr++ = stur64_offset(base, data_reg, base_offset);
+            EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset, data_reg, 4);
+            uint32_t *store_long_done = ptr++;
+
+            uint32_t *store_quad_label = ptr;
+            EMU68_GUEST_STORE_OFFSET(ptr, base, base_offset, data_reg, 8);
+            uint32_t *store_done = ptr;
+            *store_not_byte = b_cc(A64_CC_NE,
+                                   store_half_label - store_not_byte);
+            *store_not_half = b_cc(A64_CC_GT,
+                                   store_long_label - store_not_half);
+            *store_not_long = b_cc(A64_CC_GT,
+                                   store_quad_label - store_not_long);
+            *store_byte_done = b(store_done - store_byte_done);
+            *store_half_done = b(store_done - store_half_done);
+            *store_long_done = b(store_done - store_long_done);
         }
     }
 
@@ -1982,7 +2009,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
         *ptr++ = orr_immed(mask_reg, 31, width, 24 + width);
 
         // Load data 
-        *ptr++ = ldrb_offset(base, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, base, tmp, 1, 0);
 
         // Shift mask to correct position
         *ptr++ = lsrv(mask_reg, mask_reg, off_reg);
@@ -2114,7 +2141,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
             if (op != OP_EXTS && op != OP_EXTU && op != OP_FFO)
             {
                 // Store back
-                *ptr++ = strb_offset(base, tmp, 0);
+                EMU68_GUEST_STORE(ptr, base, tmp, 1);
             }
         }
     }
@@ -2124,7 +2151,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
         *ptr++ = orr_immed(mask_reg, 31, width, 16 + width);
 
         // Load data 
-        *ptr++ = ldrh_offset(base, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, base, tmp, 2, 0);
 
         if (op == OP_INS)
         {
@@ -2259,7 +2286,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
             if (op != OP_EXTS && op != OP_EXTU && op != OP_FFO)
             {
                 // Store back
-                *ptr++ = strh_offset(base, tmp, 0);
+                EMU68_GUEST_STORE(ptr, base, tmp, 2);
             }
         }
     }
@@ -2269,7 +2296,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
         *ptr++ = orr_immed(mask_reg, 31, width, width);
 
         // Load data 
-        *ptr++ = ldr_offset(base, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, base, tmp, 4, 0);
 
         if (op == OP_INS)
         {
@@ -2401,7 +2428,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
             if (op != OP_EXTS && op != OP_EXTU && op != OP_FFO)
             {
                 // Store back
-                *ptr++ = str_offset(base, tmp, 0);
+                EMU68_GUEST_STORE(ptr, base, tmp, 4);
             }
         }
     }
@@ -2411,7 +2438,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
         *ptr++ = orr64_immed(mask_reg, 31, width, width, 1);
 
         // Load data and shift it left according to reminder in offset reg
-        *ptr++ = ldr64_offset(base, tmp, 0);
+        EMU68_GUEST_LOAD(ptr, base, tmp, 8, 0);
 
         if (op == OP_INS)
         {
@@ -2550,7 +2577,7 @@ static inline uint32_t *EMIT_BFxxx_RI(uint32_t *ptr, uint8_t base, enum BF_OP op
             if (op != OP_EXTS && op != OP_EXTU && op != OP_FFO)
             {
                 // Store back
-                *ptr++ = str64_offset(base, tmp, 0);
+                EMU68_GUEST_STORE(ptr, base, tmp, 8);
             }
         }
     }
@@ -2601,26 +2628,36 @@ static inline uint32_t *EMIT_BFxxx_RR(uint32_t *ptr, uint8_t base, enum BF_OP op
     *ptr++ = add_reg(base, base_orig, off_reg_orig, ASR, 3);
     *ptr++ = and_immed(off_reg, off_reg_orig, 3, 0);
     
-    /* Fetch the data */
-    /* Width == 1? Fetch byte */
+    /* Fetch the smallest container that covers the runtime width. */
     *ptr++ = cmp_immed(width_reg, 1);
-    *ptr++ = b_cc(A64_CC_NE, 4);
-    *ptr++ = ldrb_offset(base, data_reg, 0);
+    uint32_t *not_byte = ptr++;
+    EMU68_GUEST_LOAD(ptr, base, data_reg, 1, 0);
     *ptr++ = ror64(data_reg, data_reg, 8);
-    *ptr++ = b(12);
-    /* Width <= 8? Fetch half word */
+    uint32_t *byte_done = ptr++;
+
+    uint32_t *half_label = ptr;
     *ptr++ = cmp_immed(width_reg, 8);
-    *ptr++ = b_cc(A64_CC_GT, 4);
-    *ptr++ = ldrh_offset(base, data_reg, 0);
+    uint32_t *not_half = ptr++;
+    EMU68_GUEST_LOAD(ptr, base, data_reg, 2, 0);
     *ptr++ = ror64(data_reg, data_reg, 16);
-    *ptr++ = b(7);
-    /* Width <= 24? Fetch long word */
+    uint32_t *half_done = ptr++;
+
+    uint32_t *long_label = ptr;
     *ptr++ = cmp_immed(width_reg, 24);
-    *ptr++ = b_cc(A64_CC_GT, 4);
-    *ptr++ = ldr_offset(base, data_reg, 0);
+    uint32_t *not_long = ptr++;
+    EMU68_GUEST_LOAD(ptr, base, data_reg, 4, 0);
     *ptr++ = ror64(data_reg, data_reg, 32);
-    *ptr++ = b(2);
-    *ptr++ = ldr64_offset(base, data_reg, 0);
+    uint32_t *long_done = ptr++;
+
+    uint32_t *quad_label = ptr;
+    EMU68_GUEST_LOAD(ptr, base, data_reg, 8, 0);
+    uint32_t *load_done = ptr;
+    *not_byte = b_cc(A64_CC_NE, half_label - not_byte);
+    *not_half = b_cc(A64_CC_GT, long_label - not_half);
+    *not_long = b_cc(A64_CC_GT, quad_label - not_long);
+    *byte_done = b(load_done - byte_done);
+    *half_done = b(load_done - half_done);
+    *long_done = b(load_done - long_done);
 
     /* In case of INS, prepare the source data accordingly */
     if (op == OP_INS)
@@ -2736,26 +2773,39 @@ static inline uint32_t *EMIT_BFxxx_RR(uint32_t *ptr, uint8_t base, enum BF_OP op
 
         if (op != OP_EXTS && op != OP_EXTU && op != OP_FFO)
         {
-            /* Store the data back */
-            /* Width == 1? Fetch byte */
+            /* Store the selected container back. */
             *ptr++ = cmp_immed(width_reg, 1);
-            *ptr++ = b_cc(A64_CC_NE, 4);
+            uint32_t *store_not_byte = ptr++;
             *ptr++ = ror64(data_reg, data_reg, 64 - 8);
-            *ptr++ = strb_offset(base, data_reg, 0);
-            *ptr++ = b(12);
-            /* Width <= 8? Fetch half word */
+            EMU68_GUEST_STORE(ptr, base, data_reg, 1);
+            uint32_t *store_byte_done = ptr++;
+
+            uint32_t *store_half_label = ptr;
             *ptr++ = cmp_immed(width_reg, 8);
-            *ptr++ = b_cc(A64_CC_GT, 4);
+            uint32_t *store_not_half = ptr++;
             *ptr++ = ror64(data_reg, data_reg, 64 - 16);
-            *ptr++ = strh_offset(base, data_reg, 0);
-            *ptr++ = b(7);
-            /* Width <= 24? Fetch long word */
+            EMU68_GUEST_STORE(ptr, base, data_reg, 2);
+            uint32_t *store_half_done = ptr++;
+
+            uint32_t *store_long_label = ptr;
             *ptr++ = cmp_immed(width_reg, 24);
-            *ptr++ = b_cc(A64_CC_GT, 4);
+            uint32_t *store_not_long = ptr++;
             *ptr++ = ror64(data_reg, data_reg, 32);
-            *ptr++ = str_offset(base, data_reg, 0);
-            *ptr++ = b(2);
-            *ptr++ = str64_offset(base, data_reg, 0);
+            EMU68_GUEST_STORE(ptr, base, data_reg, 4);
+            uint32_t *store_long_done = ptr++;
+
+            uint32_t *store_quad_label = ptr;
+            EMU68_GUEST_STORE(ptr, base, data_reg, 8);
+            uint32_t *store_done = ptr;
+            *store_not_byte = b_cc(A64_CC_NE,
+                                   store_half_label - store_not_byte);
+            *store_not_half = b_cc(A64_CC_GT,
+                                   store_long_label - store_not_half);
+            *store_not_long = b_cc(A64_CC_GT,
+                                   store_quad_label - store_not_long);
+            *store_byte_done = b(store_done - store_byte_done);
+            *store_half_done = b(store_done - store_half_done);
+            *store_long_done = b(store_done - store_long_done);
         }
     }
 
@@ -5420,4 +5470,4 @@ int M68K_GetLineELength(uint16_t *insn_stream)
     }
 
     return length;
-}
\ No newline at end of file
+}
